AndroidSDK/src/com/leanplum/internal/Request.java

```diff
@@ -530,51 +591,40 @@ public void sendEventually() {
    * Essentially we drop every <code>start(B)</code> call, that is directly followed by any kind of
    * a <code>start</code> call.
    *
-   * @param requestData A list of the requests, stored on the device.
-   * @return A list of only these requests, which contain relevant information for the API call.
+   * @param requestData A Map of the requests, stored on the device.
+   * @return A Map of only these requests, which contain relevant information for the API call.
    */
-  private static List<Map<String, Object>> removeIrrelevantBackgroundStartRequests(
-      List<Map<String, Object>> requestData) {
-    List<Map<String, Object>> relevantRequests = new ArrayList<>();
-
+  private static Map<Integer, Map<String, Object>> removeIrrelevantBackgroundStartRequests(
+      Map<Integer, Map<String, Object>> requestData) {
+    Map<Integer, Map<String, Object>> relevantRequests = new TreeMap<>();
     int requestCount = requestData.size();
     if (requestCount > 0) {
+      List<Integer> keysList = new ArrayList<>(requestData.keySet());
       for (int i = 0; i < requestCount; i++) {
-        Map<String, Object> currentRequest = requestData.get(i);
+        int key = keysList.get(i);
+        Map<String, Object> currentRequest = requestData.get(key);
         if (i < requestCount - 1
-            && Constants.Methods.START.equals(requestData.get(i + 1).get(Constants.Params.ACTION))
+            && Constants.Methods.START.equals(
+            requestData.get(keysList.get(i + 1)).get(Constants.Params.ACTION))
             && Constants.Methods.START.equals(currentRequest.get(Constants.Params.ACTION))
             && Boolean.TRUE.toString().equals(currentRequest.get(Constants.Params.BACKGROUND))) {
           continue;
         }
-        relevantRequests.add(currentRequest);
+        relevantRequests.put(key, currentRequest);
       }
     }
 
     return relevantRequests;
   }
 
-  private static String jsonEncodeUnsentRequests(List<Map<String, Object>> requestData) {
+  private static String jsonEncodeUnsentRequests(Map<Integer, Map<String, Object>> requestData) {
     Map<String, Object> data = new HashMap<>();
-    data.put(Constants.Params.DATA, requestData);
-    return JsonConverter.toJson(data);
-  }
-
-  private static void pushUnsentRequests(List<Map<String, Object>> requestData) {
-    if (requestData == null) {
-      return;
-    }
-    for (Map<String, Object> args : requestData) {
-      Object retryCountString = args.get("retryCount");
-      int retryCount;
-      if (retryCountString != null) {
-        retryCount = Integer.parseInt(retryCountString.toString()) + 1;
-      } else {
-        retryCount = 1;
-      }
-      args.put("retryCount", Integer.toString(retryCount));
-      saveRequestForLater(args);
+    List<Map<String, Object>> allData = new ArrayList<>();
+    for (Integer key : requestData.keySet()) {
```

Comment:
looks like we have to loop over all requests every time on encode, prob. on decode again? can we avoid the loop somehow?